### DISCUSSION TITLE
Serialize/Deserialize json to/from Rule (avoiding RuleBuilder)

### DISCUSSION THREAD

**Discussion by asulwer** [NRules/NRules#363] [Q&A] [answered]
sorry in advance for the title

using RuleBuilder is super cumbersome, in my opinion.  all the examples use a Rule class and RuleBuilder is barely touched.  a super brief example of RuleBuilder exists, if it can be called an example. Json example uses RulesBuilder so good luck with that.

What i did below works BUT is it the best way?

Rule (unrelated, is the line ctx.Update needed in the Rule?)
```
public class TestRule : Rule
{
    public override void Define()
    {
        Card card = default!;

        When()
            .Match(() => card, c => c.Name.Contains(" - FOIL"));

        Then()
            .Do(ctx => card.Replace(" - FOIL", string.Empty))
            .Do(ctx => ctx.Update(card));
    }
}
```

Card class, example object to test code
```
public static class CardExtensions
{
    public static void Replace(this Card card, string oldValue, string newValue)
    {
        card.Name = card.Name.Replace(oldValue, newValue);
    }
}

public class Card : IEquatable<Card>
{
    [Key]
    public Guid Id { get; } = Guid.NewGuid();
    public string Name { get; set; } = string.Empty;
    public string Site { get; set; } = string.Empty;
    public string Extra { get; set; } = string.Empty;

    public bool Equals(Card? c) => Id == c?.Id;
    public override bool Equals(object? obj) => Equals(obj as Card);
    public override int GetHashCode() => (Id).GetHashCode();        
}
```

Example: Serializing Rule object to json / Deserialize json to object
```
var cards = new List<Card>
{
    new Card { Name = "Test1 - FOIL", Site = "Abu" },
    new Card { Name = "Test2 - FOIL", Site = "Abu" },
    new Card { Name = "Test3 - FOIL", Site = "Abu" },
    new Card { Name = "Test1 - Foil", Site = "Abu" },
    new Card { Name = "Test2 - Foil", Site = "Abu" },
    new Card { Name = "Test3 - Foil", Site = "Abu" },
    new Card { Name = "Test1 (Borderless)", Site = "Abu" },
    new Card { Name = "Test2 (Borderless)", Site = "Abu" },
    new Card { Name = "Test3 (Borderless)", Site = "Abu" }
};

var repository = new RuleRepository();
repository.Load(x => x.From(typeof(TestRule)));

// uncomment the block below.  one line needs to be commented out
/*
var typeResolver = new TypeResolver();
typeResolver.RegisterDefaultAliases();
typeResolver.RegisterAlias("Card", typeof(Card));
typeResolver.RegisterAlias("Context", typeof(IContext));

var options = new JsonSerializerOptions
{
    WriteIndented = true,
    PropertyNamingPolicy = JsonNamingPolicy.CamelCase
};
RuleSerializer.Setup(options, typeResolver);

var rule = repository.GetRules();

var json = JsonSerializer.Serialize(rule, options);
var rules = JsonSerializer.Deserialize<IRuleDefinition[]>(json, options)!;

var compiler = new RuleCompiler();
var factory = compiler.Compile(rules);
*/

var factory = repository.Compile(); //comment this IF the previous lines are uncommented
var session = factory.CreateSession();

session.InsertAll(cards);
session.Fire();
```

**Comment by snikolayev** [comment]
@asulwer I converted the issue you created to a discussion, as I would like to keep issues focused on tracking bugs and enhancements.
You are asking whether your example is the best way, but TBH I'm not clear on what exactly you are trying to achieve; can you please elaborate? You mentioned RuleBuilder, but what are you trying to use it for? 

I also in case this is helpful, take a look at the samples:
* RuleBuilder: https://github.com/NRules/NRules/tree/develop/samples/RuleBuilder
* Rules serialization: https://github.com/NRules/NRules/tree/develop/samples/JsonRules

**Reply by asulwer** [reply]
thank you for changing to a discussion, this is the best place it as it is not an issue.  really it's about how i am using.

**Comment by asulwer** [comment]
@snikolayev creating and understanding that Rule is best when a class inherits from Rule.  using RuleBuilder to create the rule is not intuitive.  the goal is to deserialize any rules that i use, stored as json in a db.  how they are generated was in question.

**Reply by snikolayev** [reply, marked as answer]
RuleBuilder is not intended for people to be writing concrete rules. It is intended for DSLs and translators to use to generate rules from some other form. For example, JSON serialization uses it when translating rules from JSON format to the rule model.
You certainly can write rules using fluent DSL (inheriting from the Rule class) and then serialize them into JSON, so that you can store them in the DB. But if the DSL is the source of truth, you could also distribute the rules as DLLs. 
If you intend to use JSON as the source of truth, and are going to edit JSON directly, you may as well just serialize a couple of sample rules to get a sense of the JSON rule structure, and just write rules in JSON.
If JSON seems too complicated, and you are looking for a simple textual DSL, check out https://github.com/NRules/NRules.Language if that fits the bill.

**Reply by asulwer** [reply]
i do not want to recompile my code just to change a rule

**Reply by asulwer** [reply]
NRules.Language is exactly what i wanted!
